### ISSUE TITLE
Inconsistent array behaviour when passed to a function

### ISSUE THREAD

**Issue by MilanVasko** [godotengine/godot#2021]
I noticed a rather strange behaviour when dealing with RawArrays (or Arrays, for that matter). Following [the wiki](https://github.com/okamstudio/godot/wiki/tutorial_gdscript_efficiently#pointers--referencing), I wrote some code modifying a RawArray thinking it would affect outside code, since arrays are said to be passed by reference. I found this is not the case:

This is the code I was testing:
```gdscript
static func set_uint32(buff, value):
   print("in function:")
   print(buff.get(0), ", ", buff.get(1), ", ", buff.get(2), ", ", buff.get(3))

   buff.set(0, (value >> 24) & 0xFF)
   buff.set(1, (value >> 16) & 0xFF)
   buff.set(2, (value >>  8) & 0xFF)
   buff.set(3, (value >>  0) & 0xFF)

   print(buff.get(0), ", ", buff.get(1), ", ", buff.get(2), ", ", buff.get(3))
   print("end of function")

func some_important_function():
   # starts here
   var arr = RawArray()
   for i in range(4):
      arr.push_back(255)

   print("outside:")
   print(arr.get(0), ", ", arr.get(1), ", ", arr.get(2), ", ", arr.get(3))
   set_uint32(arr, 500)
   print("outside again:")
   print(arr.get(0), ", ", arr.get(1), ", ", arr.get(2), ", ", arr.get(3))
```

...which produced this output:
```
outside:
255, 255, 255, 255
in function:
255, 255, 255, 255
0, 0, 1, 244
end of function
outside again:
255, 255, 255, 255
```
This shows the array remained unmodified after the function was run.

As [Brice pointed out](http://www.godotengine.org/forum/viewtopic.php?f=12&t=2062#p9795) in [the thread I started](http://www.godotengine.org/forum/viewtopic.php?f=12&t=2062), the code above works if you swap the arr declaration from `var arr = RawArray()` to `var arr = []` and adapt the code accordingly. If you use `var arr = Array()` instead, you're back at the original behaviour.

**Comment by mikica1986vee** [comment]
It's by design, I know I read it on the wiki (or was it in the help tab :/ ), but I can't find it now.

**Comment by reduz** [comment]
yeah, all the typed arrays are atomic, this will change eventually

**Comment by mikica1986vee** [comment]
Will we have a way to replicate current behavior?

**Comment by akien-mga** [comment]
Would #2108 be the same issue? (if so I'll close it as a duplicate of this one (#2021) since it is IMO clearer).

**Comment by neikeq** [comment]
I think this could be closed as a duplicate of #5277, since the last contains enough information and has a broad discussion on the topic.

**Comment by akien-mga** [comment]
Agreed. Closing as duplicate of #5277.
